### DISCUSSION TITLE
One problem about MPI and CUDA

### DISCUSSION THREAD

**Discussion by lijiangdong1999** [BLAST-WarpX/warpx#4971] [Installation]
Hello, I use spack to install warpx with mpi and cuda in my cluster. I want to run it with some gpus, but I find it can only run many ntasks with one gpu. 
Followed is my problem.
You can see if ntask is such as 4, then one gpu is used with 4 ntasks. 
![image](https://github.com/ECP-WarpX/WarpX/assets/114063761/dd49edd2-47c7-41f8-abd1-d960f6e8420a)
![image](https://github.com/ECP-WarpX/WarpX/assets/114063761/c6c417ce-67e3-4345-b961-1cf88dcc99e1)
If ntask is 1, then one gpu is used with 1 ntask. 
![image](https://github.com/ECP-WarpX/WarpX/assets/114063761/b3b2f795-9b70-4a7d-b0f3-918600657aa7)
![image](https://github.com/ECP-WarpX/WarpX/assets/114063761/978141ef-e6d3-41cc-b86c-0999d1cfc0e9)
It's very strange.
What I want is that one node run many gpus, and one gpu run one ntask. Do you have some solution? Thanks in advance.

**Comment by ax3l** [comment]
Hi @lijiangdong1999 , this looks like a _slurm pinning issue_, as if your MPI processes are all assigned the same GPU.

It is best to coordinate with your cluster admins to get a 1-MPI-rank to 1-exclusive-GPU pinning. Try for instance this with Slurm:
```bash
# ...
#SBATCH --exclusive
#SBATCH --ntasks-per-node=8
#SBATCH --gpus-per-node=8
# ...
```
instead of your `--gres=gpu:N` flag. Does that help?

Otherwise, can you share more details, e.g., the documentation page of the cluster you are on?

**Reply by lijiangdong1999** [reply]
I am sorry, it doesn't help. The cluster I use is built by my colleague. I can only give you some information about my cluster.

**Comment by lijiangdong1999** [comment]
![image](https://github.com/ECP-WarpX/WarpX/assets/114063761/52c598df-b15f-4206-bb72-281af3b4762f)
I am sorry, it doesn't help. The cluster I use is built by my colleague. I can only give you some information about my cluster.
[server_info.txt](https://github.com/user-attachments/files/15801108/server_info.txt)
![a3d060c578473985fd91619c9f2275a](https://github.com/ECP-WarpX/WarpX/assets/114063761/a7c2cb71-70c6-4ff7-a135-92fbe0be43b6)

**Reply by lijiangdong1999** [reply]
@ax3l This is the information about my cluster. Thanks in advance.
